fiat-web/src/main/java/com/netflix/spinnaker/fiat/config/DefaultResourcePermissionConfig.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2019 Google, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.fiat.config;
+
+import com.netflix.spinnaker.fiat.model.resources.Account;
+import com.netflix.spinnaker.fiat.model.resources.Application;
+import com.netflix.spinnaker.fiat.model.resources.BuildService;
+import com.netflix.spinnaker.fiat.providers.*;
+import java.util.List;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+class DefaultResourcePermissionConfig {
```

Comment:
I don't think `AggregatingResourcePermissionProvider` being the default is a good plan. I think it's going to come back to bite us in the future. This policy is essentially "every `PermissionSource` that is defined in the code base should be turned on all at the same time". It works for now (since there's only one!) but makes no sense at all when we add more and more of these.

I understand it probably makes things easier for you, but IMO if users want something besides "permissions stored in front50", they should have to manually specify what they want in a config file somewhere.

Reply:
In 92d1359 I added a DefaultResourcePermissionProvider that takes a single source only, and made that the default. Arguably that could have been AggregatingResourcePermissionProvider with a single entry list, but DefaultRPP is a simple enough class and keeps the intent more obvious.

I left aggregation in as a configuration time switch - the group based permission model and an internal integration I'm working on would both leverage this approach, so that feels like enough use-cases to warrant it being in the codebase to me.

Reply:
I guess what concerns me about is that now adding a new `PermissionSource` is potentially a breaking change... you're essentially constraining all future `PermissionSources` to act in a way that won't screw things up, even if they have nothing at all to do with you. And you're relying on future authors to even know that any new `PermissionSource` needs to safely interact with _all_ the others (including ones that Netflix has added, but which aren't in the OSS build).

This doesn't seem like a huge thing for Netflix to deal with, but I don't think it belongs in the OSS build.

Reply:
But sources are not really interacting with each other. They are receiving a resource with well-defined permissions, and producing a resource with well-defined permission. The `ResourcePermissionProvider` is the one deciding the interactions.

In my group example, the aggregator works well whether I provide my whole group logic as one `Source`, or provide each of my groups as a separate `Source`. In both of those cases, the sources are completely isolated from the way the interaction is managed.

Reply:
Sorry, yeah, I like the source/provider distinction that Cam made, I think it's cool. My concern here is specifically with the `AggregatingResourcePermissionProvider`. That one aggregates from every single `ResourcePermissionSource` of a given type that's defined anywhere in the classpath. If someone wants to add a new permission source, it gets lumped in with all the others even though there's no way to tell that's going to happen. That's my concern and why I specifically don't think `AggregatingResourcePermissionProvider` belongs in the code base.

Reply:
I see what you mean, but I still think an aggregating provider is a general-enough usecase, because `AggregatingResourcePermissionProvider` works great with `deny-all` models, that start by denying all privileges from users, and then adding them incrementally

Reply:
But what if we want to use a different model? What if I want to check in a `AllowByDefaultResourcePermissionProvider` that uses a new `ForbiddenResourcePermissionSource`. That provider is subtractive, rather than additive.

Is that something we'd want to do? I don't know! But as this stands there is _no way_ to check in a `ResourcePermissionSource` like that without breaking everyone who sets their permission provider to `aggregate`. You've tied our hands as to what sorts of `ResourcePermissionSources` are allowed, forever.

Reply:
I suppose the answer there is that we just have to define a different interface to use, other than `ResourcePermissionSource.`

Reply:
I'm not sure the example you're giving provides an actual usecase.

If we think of it this way: `Sources` are sources from which we read permissions of a resource, while `Providers` are the ones who read all those permissions, and decide how to consolidate them. A `ForbiddenSource` is a source trying to play the role of a provider, and trying to dictate how sources other than itself should be interpreted, which doesn't make sense.

In general, I don't think a subtractive provider really works. I think to implement an allow-by-default strategy, you need a "priority" provider, which provides full permissions if no sources exist, and chooses the highest priority source (based on some rules) otherwise

Reply:
I'm not sure the example you're giving provides an actual usecase.

If we think of it this way: Sources are sources from which we read permissions of a resource, while Providers are the ones who read all those permissions, and decide how to consolidate them. A ForbiddenSource is a source trying to play the role of a provider, and trying to dictate how sources other than itself should be interpreted, which doesn't make sense.

In general, I don't think a subtractive provider really works. I think to implement an allow-by-default strategy, you need a "priority" provider, which provides full permissions if no sources exist, and chooses the highest priority source (based on some rules) otherwise